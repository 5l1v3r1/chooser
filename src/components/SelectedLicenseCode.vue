<template>
    <div class="license-code">
        <b-tabs
            class="attribution-tab"
            v-model="activeTab"
        >
            <b-tab-item :label="this.$t('use-license.rich-text-label')">
                <div id="attribution-richtext">
                <p xmlns:dct="http://purl.org/dc/terms/" xmlns:cc="http://creativecommons.org/ns#">
                    <span v-html="this.licenseCodeTextSpan" />
                    <LicenseIcons
                        :url="this.licenseUrl"
                        :iconsArr="this.iconsList"/>
                </p>
                </div>
            </b-tab-item>
            <b-tab-item :label="this.$t('use-license.html-label')">
                <div class='control' id='generated-html-container'>
                    <textarea id='attribution-html'
                              class='textarea'
                              :value="htmlLicenseParagraph"
                              readonly
                    />
                </div>
            </b-tab-item>
            <b-tab-item
                icon-pack="fas"
                icon="copy"
            >
                <template slot="header">
                    <a
                        :class="'copyBtn'"
                        :data-clipboard-target=this.clipboardTarget()>
                        <b-icon
                            icon-pack="fas"
                            icon="copy"/>
                        <span class="button-text">{{ copyText }}
                        </span>
                    </a>
                </template>
                <div class="dummy" />
            </b-tab-item>
        </b-tabs>
    </div>
</template>
<script>
import Clipboard from 'clipboard'
import LicenseIcons from './LicenseIcons'
import { mapGetters, mapState } from 'vuex'

export default {
    name: 'SelectedLicenseCode',
    components: {
        LicenseIcons
    },
    props: ['value'],
    data() {
        return {
            success: false,
            clipboard: null,
            currentTab: 0,
            copyText: this.$t('use-license.copy-label'),
            currentSelection: 'richtext'
        }
    },
    methods: {
        onCopySuccess(e) {
            this.success = true
            if (process.env.NODE_ENV === 'production') {
                const fieldsFilled = {}
                for (const detail in this.attributionDetails) {
                    fieldsFilled[detail] = this.attributionDetails[detail] !== ''
                }
                const copiedLicense = {
                    license: this.shortName,
                    // codeType can be either rich-text or html
                    codeType: this.currentSelection,
                    fieldsFilled: fieldsFilled
                }
                this.$ga.event({
                    eventCategory: 'Attribution',
                    eventAction: 'copied',
                    eventLabel: JSON.stringify(copiedLicense)
                })
            }
            this.$emit('copied', { content: e.text })
            setTimeout(() => {
                this.success = false
            }, 2000)
            e.clearSelection()
        },
        onCopyError(e) {
            this.$emit('copyFailed')
            e.clearSelection()
        },
        clipboardTarget() {
            return `#attribution-${this.currentSelection}`
        }
    },
    computed: {
        ...mapState({ attributionDetails: state => state.attributionDetails }),
        ...mapGetters(['shortName', 'fullName', 'iconsList', 'licenseUrl']),
        licenseLink() {
            const short = this.shortName.toUpperCase()
            const attrs = 'target="_blank" rel="license noopener noreferrer" style="display: inline-block;">'
            return `<a href="${this.licenseUrl}" ${attrs}${short}</a>`
        },
        authorElement() {
            const name = this.attributionDetails.creatorName
            const profile = this.attributionDetails.creatorProfileUrl
            if (name) {
                const creatorNameNoLink = `<span rel="cc:attributionName">${name}</span>`
                return profile
                    ? `by <a rel="cc:attributionURL"  href="${profile}">${creatorNameNoLink}</a>`
                    : `by ${creatorNameNoLink}`
            } else return ''
        },
        titleElement() {
            const workUrl = this.attributionDetails.workUrl
            const workTitle = this.attributionDetails.workTitle
            if (!workTitle && !workUrl) {
                return this.$t('code-text.this-work')
            } else {
                const titleSpan = workTitle ? `<span rel="dc:title">${workTitle}</span>`
                    : this.$t('code-text.this-work')
                if (workUrl) {
                    return `<a rel="cc:attributionURL" href="${workUrl}">${titleSpan}</a>`
                } else {
                    return titleSpan
                }
            }
        },
        licenseCodeTextSpan() {
            const attributionCode = this.$t('code-text.license-text', {
                workTitle: this.titleElement,
                byLine: this.authorElement
            })
            return `${attributionCode}${this.licenseLink}`
        },
        htmlLicenseParagraph() {
            let short = this.shortName
            const iconStyle = 'style="height:22px!important;margin-left: 3px;vertical-align:text-bottom;opacity:0.7;"'
            const baseAssetsPath = 'https://search.creativecommons.org/static/img'
            let licenseIcons = `<img ${iconStyle} src="${baseAssetsPath}/cc_icon.svg" />`
            if (short.includes('CC0')) {
                short = 'CC CC0 1.0'
            }
            licenseIcons += short.slice(3, short.length - 4).split('-').map(license =>
                `<img  ${iconStyle} src="${baseAssetsPath}/cc-${license.toLowerCase()}_icon.svg" />`
            ).join('')
            const licenseIconsLink = `<a href="${this.licenseURL}">${licenseIcons}</a>`
            const ccrelcode = 'xmlns:dct="http://purl.org/dc/terms/" xmlns:cc="http://creativecommons.org/ns#"'
            return `<p ${ccrelcode}>${this.licenseCodeTextSpan}${licenseIconsLink}</p>`
        },
        activeTab: {
            get() { return this.currentTab },
            set(val) {
                if (val !== 2) {
                    this.currentTab = val
                } else {
                    this.currentSelection = this.currentTab === 0 ? 'richtext' : 'html'
                    const tab = this.currentTab
                    this.currentTab = 2
                    this.copyText = this.$t('use-license.copied-label')
                    setTimeout(() => {
                        this.currentTab = tab
                    }, 1)
                    setTimeout(() => {
                        this.copyText = 'Copy'
                    }, 2000)
                }
            }
        }
    },
    mounted() {
        this.clipboard = new Clipboard('.copyBtn')
        this.clipboard.on('success', this.onCopySuccess)
        this.clipboard.on('error', this.onCopyError)
    },
    destroyed() {
        this.clipboard.destroy()
    }
}
</script>
<style lang="scss">
    #attribution-richtext p {
        margin-top: 1rem;
        margin-bottom: 1rem;
    }
    .use-license-hint {
        text-align: center;
    }
    .use-license-hint a {
        text-decoration: underline;
    }
    .use-license-hint span {
        vertical-align: middle;
    }
    .license-code {
        margin-top: 0.4rem;
        div.attribution-tab {
            margin-bottom: 1rem;
            nav.tabs {
                margin-bottom: 0;
                ul {
                    li {
                        & > a {
                            & > span {
                                    font-style: normal;
                                    font-weight: 700;
                                    font-size: 12px;
                                    line-height: 18px;
                                    text-transform: uppercase;
                                    letter-spacing: 0.5px;
                                    color: #333333;
                                    opacity:0.8;

                            }
                        }
                        &:last-of-type{
                            margin-left: auto;
                            margin-bottom: -1px;
                            a {
                                padding:0;
                            }
                            & > a {
                                padding: 0.3em 1em;
                                span a.copyBtn {
                                    border-bottom: none;
                                }
                            }
                        }
                        &.is-active {
                            a {
                                border-bottom-width: 3px;
                                span{
                                    font-weight: 700;
                                    opacity:1;
                                }
                            }
                        }
                    }
                }
            }
            section.tab-content {
                margin-top:0;
                padding: 0;
                div.tab-item {
                    height: 60px;
                    div#generated-html-container {
                        padding-top:10px;
                        padding-bottom: 10px;
                        textarea {
                            word-break: break-all;
                            -ms-word-break: break-all;
                            min-height: 60px;
                            font-family: "Source Sans Pro", "Noto Sans", Arial, "Helvetica Neue", Helvetica, sans-serif;
                            font-size:0.845rem;
                        }
                    }
                    div#generated-richtext-container {
                        margin-top: 1rem;
                        display: block;
                        .attribution-license-icons {
                            vertical-align: middle;
                            margin-top: 0;
                            margin-left: 2px;
                        }
                        p {
                            display: inline
                        }
                    }
                }
            }
        }
    }

    @media only screen and (max-width: 1024px) {
        .card {
            height: auto!important;
        }
    }
</style>
