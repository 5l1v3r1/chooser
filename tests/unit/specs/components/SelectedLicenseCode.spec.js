import { mount, createLocalVue, config } from '@vue/test-utils'
import Buefy from 'buefy'
import SelectedLicenseCode from '@/components/SelectedLicenseCode'
import LicenseUtilities from '@/utils/license-utilities'
import Vue from 'vue'
import VueI18n from 'vue-i18n'

const value = {
    shortName: 'CC BY 4.0',
    fullName: 'Attribution 4.0 International',
    attributionDetails: {
        creatorName: '',
        creatorProfileUrl: '',
        workTitle: '',
        workUrl: ''
    }
}

// mock dom methods
function _mockDomMethodsForClipboardJS(value) {
    window.getSelection = () => ({
        addRange: () => {},
        removeAllRanges: () => {},
        toString: () => value
    })

    document.execCommand = () => value
    document.createRange = () => ({ selectNodeContents: () => { } })
}

describe('SelectedLicenseCode.vue', () => {
    let wrapper

    beforeEach(() => {
        const localVue = createLocalVue()
        localVue.use(Buefy)
        localVue.use(LicenseUtilities)
        Vue.use(VueI18n)
        const messages = require('@/locales/en.json')
        const i18n = new VueI18n({
            locale: 'en',
            fallbackLocale: 'en',
            messages: messages
        })

        config.mocks.i18n = i18n

        config.mocks.$t = (key) => {
            return i18n.messages[key]
        }

        _mockDomMethodsForClipboardJS()
        wrapper = mount(SelectedLicenseCode, {
            localVue,
            propsData: {
                value: value
            },
            attachToDocument: true
        })
    })
    // Snapshot tests
    it('has the expected UI', () => {
        expect(wrapper).toMatchSnapshot()
    })

    // Tabs exist
    it('has html, richtext and copy tabs', () => {
        const tabs = wrapper.findAll('li')
        expect(tabs.length).toEqual(3)
        expect(tabs.at(0).classes()).toContain('is-active')
    })

    // Tabs change
    it('Changes tabs on click', () => {
        const tabsNav = wrapper.findAll('li')
        // The tab label should equal 'Rich Text', but isn't because of i18n handling
        // TODO: find a better way to test labels
        expect(tabsNav.at(0).text()).toEqual('')
        expect(wrapper.findAll('.is-active')).toHaveLength(1)
        expect(tabsNav.at(0).classes()).toContain('is-active')

        const tabs = wrapper.findAll('.tab-item')
        expect(tabs).toHaveLength(3)
        expect(tabs.at(0).isVisible()).toBe(true)
        expect(tabs.at(1).isVisible()).toBe(false)

        tabsNav.at(1).find('a').trigger('click')
        expect(tabsNav.at(1).classes()).toContain('is-active')
        expect(wrapper.findAll('.is-active')).toHaveLength(1)
        expect(tabs.at(1).isVisible()).toBe(true)

        const copyBtn = tabsNav.at(2).find('a').find('span').find('a')

        copyBtn.trigger('click')
        // The tab label should equal 'Copied!', but isn't because of i18n handling
        expect(tabsNav.at(2).text()).toEqual('')
        setTimeout(() => {
            expect(wrapper.vm.currentTab).toEqual(1)
        }, 2000)
    })

    // Test copying
    it('Copies text from active tab', () => {
        const copyLink = wrapper.findAll('li').at(2).find('a').find('span').find('a')

        expect(wrapper.vm.currentTab).toEqual(0)
        expect(wrapper.vm.clipboardTarget()).toEqual('#attribution-richtext')
        const htmlTab = wrapper.findAll('li').at(1).find('a')
        htmlTab.trigger('click')
        copyLink.trigger('click')
        expect(wrapper.vm.clipboardTarget()).toEqual('#attribution-html')
        expect(wrapper.vm.currentSelection).toEqual('html')

        const emittedEvents = wrapper.emitted()
        expect(emittedEvents).toHaveProperty('copyFailed')
        expect(emittedEvents.copyFailed.length).toBe(1)
    })

    // Test generated HTML
    it('Generates correct license HTML', () => {
        const generatedHtml = wrapper.find('#attribution-richtext')
        expect(generatedHtml.text()).toContain('CC BY 4.0')
        // check that icon links are correct
        // There are two icons - CC & BY
        expect(generatedHtml.findAll('img')).toHaveLength(2)
        // There are three links: two icons and license name
        expect(generatedHtml.findAll('a')).toHaveLength(1)
        expect(generatedHtml.find('a').attributes('href')).toEqual('https://creativecommons.org/licenses/by/4.0/?ref=ccchooser')
        // check that the text is correct
        // check that changing fields generate correct texts
    })
})
