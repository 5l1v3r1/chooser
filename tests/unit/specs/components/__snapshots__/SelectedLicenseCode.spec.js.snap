// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedLicenseCode.vue has the expected UI 1`] = `
<div class="license-code">
  <h2 class="vocab-h2">
    Use your license</h2>
  <div class="b-tabs attribution-tab">
    <nav class="tabs">
      <ul>
        <li class="is-active"><a>
            <!----> <span>Rich Text</span></a></li>
        <li class=""><a>
            <!----> <span>HTML</span></a></li>
        <li class=""><a><span><a data-clipboard-target="#attribution-richtext" class="copyBtn"><span class="icon" icon-pack="fas"><i class="mdi mdi-copy mdi-24px"></i></span> <span class="button-text">Copy
                    </span></a></span></a></li>
      </ul>
    </nav>
    <section class="tab-content">
      <div class="tab-item" style="">
        <div id="attribution-richtext">
          <p id="license-code" xmlns:dct="http://purl.org/dc/terms/" xmlns:cc="http://creativecommons.org/ns#">{workTitle} {byLine} is licensed under <a rel="license" href="https://creativecommons.org/licenses/by/4.0/?ref=ccchooser" target="_blank" style="display: inline-block;">CC BY 4.0<span class="license-icons" style="opacity:.7"><img style="height: 22px!important;margin-left: 3px;vertical-align:text-bottom;" src="https://search.creativecommons.org/static/img/cc_icon.svg"><img style="margin-top: 2px;height: 22px!important;margin-left: 3px;vertical-align:text-bottom;" src="https://search.creativecommons.org/static/img/cc-by_icon.svg"></span></a></p>
        </div>
      </div>
      <div class="tab-item" style="display: none;">
        <div id="generated-html-container" class="control"><textarea id="attribution-html" readonly="readonly" class="textarea"></textarea></div>
      </div>
      <div class="tab-item" style="display: none;">
        <div class="dummy"></div>
      </div>
    </section>
  </div>
  <p class="use-license-hint"><span class="icon is-medium"><i class="fas fa-question-circle fa-2x"></i></span> <a href="https://wiki.creativecommons.org/wiki/Marking_your_work_with_a_CC_license">
      Learn more about how to use your license
    </a></p>
</div>
`;
